node/src/exchange/client.rs

```diff
@@ -56,63 +96,697 @@ impl ExchangeClientHandler {
+                    respond_to.maybe_send_ok(vec![resp.to_owned()]);
+                    return;
+                }
+            }
+
+            // Otherwise return the header with the maxium height
+            respond_to.maybe_send_ok(vec![resps[0].to_owned()]);
+        });
     }
 
     #[instrument(level = "trace", skip(self, responses), fields(responses.len = responses.len()))]
     pub(super) fn on_response_received(
         &mut self,
-        request_id: RequestId,
+        peer: PeerId,
+        request_id: R::Id,
         responses: Vec<HeaderResponse>,
     ) {
-        let Some(ReqInfo { amount, respond_to }) = self.reqs.remove(&request_id) else {
+        let Some(state) = self.reqs.remove(&request_id) else {
             return;
         };
 
-        trace!("Response received. Expected amount = {amount}");
+        trace!(
+            "Response received. Expected amount = {}",
+            state.request.amount
+        );
+
+        match self.decode_and_verify_responses(&state.request, &responses) {
+            Ok(headers) => {
+                // TODO: Increase peer score
+                state.respond_to.maybe_send_ok(headers);
+            }
+            Err(e) => {
+                // TODO: Decrease peer score
+                state.respond_to.maybe_send_err(e);
+            }
+        }
+    }
+
+    fn decode_and_verify_responses(
+        &mut self,
+        request: &HeaderRequest,
+        responses: &[HeaderResponse],
+    ) -> Result<Vec<ExtendedHeader>, ExchangeError> {
+        let amount = usize::try_from(request.amount).expect("validated in send_request");
 
         if responses.len() != amount {
-            // TODO: should we define a separate error for this case?
-            respond_to.maybe_send_err(ExchangeError::InvalidResponse);
-            return;
+            return Err(ExchangeError::InvalidResponse)?;
         }
 
         let mut headers = Vec::with_capacity(amount);
 
         for response in responses {
-            let res = match response.status_code() {
-                StatusCode::Invalid => Err(ExchangeError::InvalidResponse),
-                StatusCode::NotFound => Err(ExchangeError::HeaderNotFound),
-                StatusCode::Ok => ExtendedHeader::decode(&response.body[..])
-                    .map_err(|_| ExchangeError::InvalidResponse),
-            };
+            let header = response.to_extended_header()?;
+            trace!("Header: {header:?}");
+            headers.push(header);
+        }
+
+        headers.sort_unstable_by_key(|header| header.height());
 
-            match res {
-                Ok(header) => {
-                    trace!("Header: {header:?}");
-                    headers.push(header);
+        match (&request.data, request.amount) {
+            (_, 0) => return Err(ExchangeError::InvalidResponse),
+
+            // Allow HEAD requests to have any height in their response
+            (Some(Data::Origin(0)), _) => {}
```

Comment:
```suggestion
            // Allow HEAD requests to have any height in their response
            (Some(Data::Origin(0)), 1) => {}
```
Shouldn't we match on `1` there too? I know it's validated on sending side but just for case

Reply:
fixed